internal/controller/collection/reconciler.go

```diff
@@ -0,0 +1,277 @@
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, "cannot create observed object")
+		}
+
+		log.Debug("created observed object", "name", po.GetName())
+		refs.Insert(v1alpha1.ObservedObjectReference{Name: name})
+	}
+
+	// Remove objects that either do not exist anymore or are no match
+	oldRefs := sets.New[v1alpha1.ObservedObjectReference](collection.Status.Objects...)
+	for _, or := range oldRefs.Difference(refs).UnsortedList() {
+		o := &v1alpha2.Object{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: or.Name,
+			}}
+		if err := r.client.Delete(ctx, o); err != nil {
+			collection.Status.SetConditions(xpv1.ReconcileError(err))
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("cannot delete observed object %s", or.Name))
+		}
+	}
+	collection.Status.SetConditions(xpv1.ReconcileSuccess(), xpv1.Available())
+
+	// Update references only if more or less objects got matched since the last reconciliation.
+	if !refs.Equal(oldRefs) {
+		collection.Status.Objects = refs.UnsortedList()
+	}
+
+	return ctrl.Result{RequeueAfter: r.pollInterval()}, r.client.Status().Update(ctx, collection)
+}
+
+func observedObjectName(collection client.Object, matchedObject client.Object) (string, error) {
+	h := md5.New() //#nosec G401 -- used only for unique name generation
+	if _, err := fmt.Fprintf(h, "%s-%s", collection.GetName(), matchedObject.GetUID()); err != nil {
+		return "", err
+	}
+	id, err := uuid.FromBytes(h.Sum(nil))
+	return id.String(), err
+}
+
+func observedObjectPatch(name string, matchedObject unstructured.Unstructured, collection *v1alpha1.ObservedObjectCollection) (*unstructured.Unstructured, error) {
+	objectManifestTemplate := `{
+"kind": "%s",
+"apiVersion": "%s",
+"metadata": {
+  "name": "%s",
+  "namespace": "%s"
+}
+}`
+	manifest := fmt.Sprintf(objectManifestTemplate, matchedObject.GetKind(), matchedObject.GetAPIVersion(), matchedObject.GetName(), matchedObject.GetNamespace())
+	observedObject := &v1alpha2.Object{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: collection.APIVersion,
+					Kind:       collection.Kind,
+					Name:       collection.Name,
+					UID:        collection.UID,
+				},
+			},
+		},
+		Spec: v1alpha2.ObjectSpec{
+			ResourceSpec: xpv1.ResourceSpec{
+				ProviderConfigReference: &collection.Spec.ProviderConfigReference,
+				ManagementPolicies:      []xpv1.ManagementAction{xpv1.ManagementActionObserve},
+			},
+			ForProvider: v1alpha2.ObjectParameters{
+				Manifest: runtime.RawExtension{
+					Raw: []byte(manifest),
+				},
+			},
+		},
+	}
+	if t := collection.Spec.Template; t != nil {
+		if len(t.Metadata.Labels) > 0 {
+			observedObject.SetLabels(t.Metadata.Labels)
```

Comment:
By default adding a label like `kubernetes.crossplane.io/owned-by-collection: jobs` would be helpful if [not necessary](https://github.com/crossplane-contrib/provider-kubernetes/pull/217/files#r1544381370). This could be used to quickly filter those object with `kubectl get` but also in the compositions.

Reply:
If we introduce such label, then we implicitly limit collection names to 63 chars. Let me know if this is an important trade to be made. Also, `ObservedObjectCollection` instances are gonna be created mostly from within a composition, where their name is built from composite name and a prefix.

We already add possibility for user to add custom labels/annotations through `.spes.template`, do we need an additional label here that is going to limit the collection name length to 63 chars?